aten/src/ATen/functorch/BatchRulesScatterOps.cpp

```diff
@@ -1056,6 +1056,132 @@ std::tuple<Tensor,optional<int64_t>> masked_fill_scalar_batch_rule(
   return std::make_tuple(result, 0);
 }
 
+std::tuple<Tensor,optional<int64_t>> index_fill_int_scalar_batch_rule_impl(
+    Tensor & self, optional<int64_t> self_bdim,
+    int dim,
+    const Tensor & index, optional<int64_t> index_bdim,
+    const Scalar & value,
+    const bool inplace) {
+
+    const auto self_logical_rank = rankWithoutBatchDim(self, self_bdim);
+    Tensor self_ = moveBatchDimToFront(self, self_bdim);
+    Tensor index_ = moveBatchDimToFront(index, index_bdim);
+    dim = maybe_wrap_dim(dim, self_logical_rank);
+
+    if (inplace && !self_bdim.has_value()) {
+      vmapIncompatibleInplaceError("index_fill_");
+    }
+
+    if (!index_bdim) {
+      self_.unsqueeze_(-1);
+      self_.index_fill_(dim + 1, index_, value);
+      self_.squeeze_(-1);
+      return std::make_tuple(self_, 0);
+    }
+
+    auto batch_size = get_bdim_size2(self, self_bdim, index, index_bdim);
+    self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
+    index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
+
+    if (!inplace){
+      if (!self_bdim.has_value())
+        self_ = self_.clone();
+    }
+
+    for (const auto i : c10::irange(0, batch_size)) {
+      const auto& self_slice = self_.select(0, i);
+      const auto& index_slice = index_.select(0, i);
+      self_slice.index_fill_(
+        dim,
+        index_slice,
+        value
+      );
+    }
```

Comment:
For vmap we are generally trying to avoid for-loops. One can actually do this to a single call to index_fill_:
- let's say our self_ has shape (2, 3, 3) 
- we take the `index` (let's say it is [[0, 2], [1, 2]]) for convenience)
- We add some offsets to it: [[0, 2], [1 + K, 2 + K]])
- Then we put them all into a single list: [0, 2, 1 + K, 2 + K]
- If dim == 0, then we reshape self_ into (6, 3)
- and now we can apply index_fill on self_ and the new index. This results in an output of shape (6, 3)
- we can then reshape this back into (2, 3, 3) and then return it.

The reshape helper functions in https://github.com/pytorch/pytorch/blob/06bdd491fb431a7c7cc58c2402e485fcb8406ef9/aten/src/ATen/functorch/BatchRulesHelper.h#L26-L27 are designed to help out with this situation.

We should be able to apply a similar algorithm to the value is a Tensor case

Reply:
That makes sense. But for the in-place `index_fill`, wouldn't it be problematic when we reshape the `self_` having an incompatible stride, which will result in a copied tensor?

Reply:
Ahh you're right. We've got some options:
1. if `self_` has an incompatible stride... clone self into `self_clone`, do the operation out of place, and then do `self.copy_(self_clone)`. Otherwise, do the single index_fill_. This basically gets rid of the benefits of the in-place operation since we need to create an intermediate Tensor.
2. if `self_` has an incompatible stride, do the for-loop approach. Otherwise, do the single index_fill_.
3. For in-place, always do the for-loop approach. Requires less testing on our side (we don't need to come up with test cases that have incompatible/compatible strides), also the stride is likely going to be incompatible most of the time.
4. For out-of-place and in-place, always do the for-loop approach

Thoughts? The current PR (which does 4) is an improvement over the previous state, so I'm happy to accept the approach and get it merged. Though 3 seems like where we should go to squeeze out some improvements for the out-of-place index_fill case.

Reply:
I agree that option 3 could give us more improvements, though it was tricky when I was implementing it. I wrote my question in another review.